sycl/include/CL/sycl/stream.hpp

```diff
@@ -104,6 +107,15 @@ class stream {
 
   bool operator!=(const stream &LHS) const;
 
+  ~stream() {
```

Comment:
The stream class is copyable. I don't think we want to flush if a copy of the stream class is deleted. Only when the last one is deleted / the kernel finishes should we flush.

Reply:
Good question, thx. Now everything is ok if stream is copied in the host side. But it is more tricky if stream is copied in the device code, because now I use a field of the stream object to save offset in flush buffer for work item. So 2 objects (with their on offset in same flush buffer) could write to same local space. It looks like I need to use local mem for offset, for exmaple the first several bytes of each flush buffer can store offset in the flush buffer.

Reply:
It seems you fixed the more important problem that the output is correct if the stream is copied. Did you address my initial comment? It seems the stream class is still causing a flush by calling flushBuffer on each call of the destructor. In the case the destructor of a copy is called this causes an extra flush which isn't documented.

Reply:
I believe that I have addressed this problem. I do flush only if Offset is not zero, i.e. if there is some data to flush. Offset is shared between copies of the stream object. So if some of the copies flushed data then other copy will not flush because Offset will be 0.

Reply:
I agree you won't flush multiple times if all destructors are called at the same time because they all are in the same scope. But if you have:
``` c++
{ //begin of kernel
s << "a";
{ 
    auto s2 = s;
    s2 << "b";
} //destructor of s2
s << "c";
}  // end of kernel. destructor of s
```
The current implementation will cause an extra flush between b and c. The example is contrived. But if you e.g. pass the stream by value to a function you would have the same at the end of the function scope.

Reply:
This isn't blocking. But needs to be addressed at some point.

Reply:
Oh, I see, thank you for the example. Then I will address this in separate commit.